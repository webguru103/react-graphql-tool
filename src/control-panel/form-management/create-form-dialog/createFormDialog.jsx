// @flow

import * as React from 'react';
import type { Node } from 'react';
import { withDialog } from '../../../shared/dialog/withDialog';
import DialogButtonBottom from '../../../shared/dialogButtonBottom/dialogButtonBottom';
import FormInput from './formInput';
import validator from '../../../validation/helpers';
import { formValidationSchema } from '../../../validation/validationSchemas';
import { withCreateForm } from '../api/form.service';
import { pro, compose, get } from '../../../utility';
import c from '../constants';
import { errParser } from '../../../api-error-parser';
import errorMessages from '../../../api-error-parser/formattedMessages';
import type { DocumentUpload } from '../flowTypes';

type Props = {
  createForm: Function,
  closeDialog: Function,
  groupId: string,
  history: Object,
  testMode: boolean, // Used for testing to bypass file upload dialog
}

export class CreateFormDialogC extends React.PureComponent<Props, *> {

  state = {
    networkError: '',
  };

  onCreate = async ({
    name,
    status,
    documents,
  }: {
      name: string,
      status: string,
      documents: Array<DocumentUpload>,
    } = {}) => {
    let pdf = get(documents[0], 'pdf', '');
    const pages = get(documents[0], 'docPages', []);

    // TestMode added here because a file cannot be easily programmatically added to the dropzone input
    // for testing purposes. Passing in the testMode prop will add a test URL to the urls array,
    // And also bypasses the validation step in the validate function.
    if (this.props.testMode) {
      pdf = 'http://testurl.com/pdf.pdf';
    }

    if (pdf === '') {
      return;
    }

    const {
      history,
      groupId,
      createForm,
      closeDialog,
    } = this.props;

    const formInput = {
      formName: name.trim(),
      formGroupId: groupId,
      formStatus: status,
    };

    const formPages = pages.map(item => ({
      pageNumber: item.pageNumber,
      width: item.width,
      height: item.height,
    }));

    const formVersionInput = {
      sourceURL: get(documents[0], 'pdf', ''),
      formPages,
    };

    const [err, data] = await pro(createForm(formInput, formVersionInput));

    if (err) {
      const parsedErr = errParser.parse(err);
      const globalError = get(parsedErr, 'global');
      const networkError = get(parsedErr, 'network');

      // If there was a network error, just return a generic network error message.
      const errorToSet = networkError ? errorMessages.networkErrorMessage() : globalError;

      this.setState({
        networkError: errorToSet,
      });
    } else {
      history.push(`/cp-user/editor/edit?formId=${get(data, 'data.createForm.id', '')}`);
      closeDialog();
    }
  };

  render() {
    const {
      networkError,
    } = this.state;

    return (
      <DialogButtonBottom
        title={c.NEW_FORM}
        content={(
          <FormInput networkError={networkError} />
        )}
        initialValues={{
          name: '',
          status: 'ACTIVE',
          documents: [],
        }}
        validate={(values) => {
          const validationResult = validator(values, formValidationSchema);

          const errors: { documents?: Node } = {
            ...validationResult.errors,
          };

          // TestMode used here to bypass 'urls' check if test is running
          if ((!values.documents || values.documents.length < 1) && !this.props.testMode) {
            errors.documents = c.NO_FORM_BASE_UPLOADED;
          }

          return errors;
        }}
        actionButtonText={c.START_EDITING}
        onSubmit={this.onCreate}
      />
    );
  }

}

export default compose(
  withCreateForm,
  withDialog,
)(CreateFormDialogC);
