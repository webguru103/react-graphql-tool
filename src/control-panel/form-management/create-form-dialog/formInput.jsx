// @flow

import * as React from 'react';
import classNames from 'classnames';

import TextField from '@material-ui/core/TextField';
import RadioGroup from '@material-ui/core/RadioGroup';
import Radio from '@material-ui/core/Radio';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import FormControl from '@material-ui/core/FormControl';
import FormHelperText from '@material-ui/core/FormHelperText';
import FormLabel from '@material-ui/core/FormLabel';
import { withStyles } from '@material-ui/core/styles';

import UploadForm from '../../../shared/uploadForm/uploadForm';
import styles from './formInput.styles';
import { noop } from '../../../utility';
import c, { MAX_UPLOAD_SIZE, PDF } from '../constants';

type Props = {
  classes: Object,
  networkError: ?React.Node,
  values?: Object,
  touched?: Object,
  errors?: Object,
  handleChange?: Function,
  setErrors?: Function,
  setFieldValue?: Function,
};

class FormInput extends React.Component <Props, *> {

  static defaultProps = {
    values: {},
    touched: {},
    errors: {},
    handleChange: noop,
    setErrors: noop,
    setFieldValue: noop,
  };

  state = {
    uploadError: '',
  }

  setURLField = (documents: Array<any>) => {
    if (this.props.setFieldValue) {
      this.props.setFieldValue('documents', documents);
    }
  }

  setUploadError = (uploadError) => {
    this.setState({
      uploadError,
    });
  }

  render() {

    const {
      classes,
      networkError,
      values,
      touched,
      errors,
      handleChange,
      setErrors,
    } = this.props;

    const { uploadError } = this.state;

    return (
      <div className={classes.formInput}>
        <TextField
          error={touched && touched.name && errors && Boolean(errors.name)}
          fullWidth
          helperText={touched && touched.name && errors && errors.name}
          label={c.FORM_NAME}
          id="name"
          InputProps={{
            classes: {
              root: classes.textFieldInputWrapper,
              input: classes.textFieldInput,
              error: classes.textFieldInputError,
            },
            disableUnderline: true,
            inputProps: {
              'data-testid': 'form-name-input',
            },
          }}

          InputLabelProps={{
            classes: { root: `${classNames(classes.textFieldLabel, classes.label)}` },
            required: true,
          }}
          onChange={handleChange}
          value={values && values.name}
          name="name"
        />
        <FormControl>
          <FormLabel
            classes={{ root: `${classNames(classes.textFieldLabel, classes.label, classes.singleLabel)}` }}
            required
          >
            Upload Form Base
          </FormLabel>
        </FormControl>
        <UploadForm
          acceptTypes={[PDF]}
          acceptMultiple={false}
          maxUploadSize={MAX_UPLOAD_SIZE}
          uploadLabel={c.UPLOAD_LABEL}
          setUploadError={this.setUploadError}
          setURLs={this.setURLField}
          onDrop={() => {
            // Formik hook for UploadForm - reset error for urls field
            // manually so that when upload field is touched, error is no longer present
            if (setErrors) {
              setErrors({
                ...errors,
                documents: '',
              });
            }
          }}
        />
        {Boolean(uploadError) && (
          <FormControl error>
            <FormHelperText>
              {uploadError}
            </FormHelperText>
          </FormControl>
        )}
        {touched && touched.documents && Boolean(errors && errors.documents) && (
          <FormControl error>
            <FormHelperText>
              {errors && errors.documents}
            </FormHelperText>
          </FormControl>
        )}
        <FormControl component="fieldset">
          <RadioGroup
            className={classes.statusRadioGroup}
            aria-label="status"
            name="status"
            value={values && values.status}
            onChange={handleChange}
          >
            <FormControlLabel data-testid="active-form" value="ACTIVE" control={<Radio />} label="Active" />
            <FormControlLabel data-testid="inactive-form" value="INACTIVE" control={<Radio />} label="Inactive" />
          </RadioGroup>
        </FormControl>
        {Boolean(networkError) && (
          <FormControl error>
            <FormHelperText>
              {networkError}
            </FormHelperText>
          </FormControl>
        )}
      </div>
    );
  }

}

export default withStyles(styles)(FormInput);
