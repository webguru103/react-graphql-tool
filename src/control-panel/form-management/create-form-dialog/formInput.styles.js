export default (theme: Object) => ({
  formInput: {
    display: 'flex',
    flexFlow: 'column',
    width: '550px',
  },
  textInput: {
    boxShadow: 'inset 0px 1px 3px 0px',
    paddingLeft: '5px',
  },
  textFieldInputWrapper: {
    'label + &': {
      marginTop: '10px',
    },
  },
  textFieldInput: {
    padding: '10px 12px',
    border: `1px solid ${theme.palette.neutrals.greyUI}`,
    borderRadius: 4,
    color: `${theme.palette.neutrals.mainText}`,
    fontSize: 14,
    '&:focus': {
      borderColor: `${theme.palette.neutrals.mainText}`,
    },
  },
  textFieldInputError: {
    '& $textFieldInput': {
      borderColor: '#f44336',
    },
  },
  textFieldLabel: {
    position: 'relative',
    transform: 'none',
    fontSize: 14,
  },
  label: {
    color: `${theme.palette.neutrals.mainText}`,
    '& span': {
      color: `${theme.palette.neutrals.mainText}`,
    },
  },
  singleLabel: {
    margin: '20px 0 10px 0',
  },
  statusRadioGroup: {
    marginBottom: 20,
  },
  radio: {
    width: 30,
    height: 30,
  },
});
