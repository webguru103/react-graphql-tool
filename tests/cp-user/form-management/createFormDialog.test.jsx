/* eslint import/no-extraneous-dependencies: 0 */

import * as React from 'react';
import 'jest-dom/extend-expect';
import { render, cleanup, fireEvent } from 'react-testing-library';
import { withContext } from '../../test-helpers/contextWrapper';
import { CreateFormDialogC } from '../../../src/control-panel/form-management/create-form-dialog/createFormDialog';

describe('CreateFormDialogC', () => {
  afterEach(cleanup);
  const Component = withContext(CreateFormDialogC);

  test('renders', () => {
    const { container } = render(<Component
      createForm={() => {}}
      closeDialog={() => {}}
      groupId="testgroup"
      history={{}}
    />);
    expect(container).toMatchSnapshot();
  });

  test('calls createForm', (done) => {
    const createForm = jest.fn(() => Promise.resolve());

    const { getByTestId } = render(<Component
      createForm={createForm}
      closeDialog={() => {}}
      groupId="testgroup"
      history={{ push: () => {} }}
      testMode
    />);

    const nameInput = getByTestId('form-name-input');
    fireEvent.change(nameInput, { target: { value: 'changed' } });

    const submitButton = getByTestId('action-dialog-button-bottom');
    fireEvent.click(submitButton);

    setTimeout(() => {
      expect(createForm).toHaveBeenCalledTimes(1);
      done();
    });
  });

});
