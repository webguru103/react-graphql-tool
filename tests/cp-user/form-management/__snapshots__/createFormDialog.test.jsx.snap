// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateFormDialogC renders 1`] = `
<div>
  <form>
    <div
      class="MuiDialogTitle-root-7 DialogCreatorInjector-dialogTitle-2"
    >
      <span>
        <span>
          New Form
        </span>
      </span>
      <button
        class="MuiButtonBase-root-14 MuiIconButton-root-8 DialogCreatorInjector-dialogClose-6"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label-13"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-17"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      class="MuiDialogContent-root-26 DialogCreatorInjector-dialogContent-3"
    >
      <div
        class="FormInput-formInput-27"
      >
        <div
          class="MuiFormControl-root-37 MuiFormControl-fullWidth-40"
        >
          <label
            class="MuiFormLabel-root-48 MuiFormLabel-required-53 MuiInputLabel-root-41 FormInput-textFieldLabel-32 FormInput-label-33 MuiInputLabel-formControl-42 MuiInputLabel-animated-45"
            data-shrink="false"
            for="name"
          >
            <span>
              Form Name
            </span>
            <span
              class="MuiFormLabel-asterisk-54"
            >
               *
            </span>
          </label>
          <div
            class="MuiInputBase-root-68 MuiInput-root-55 FormInput-textFieldInputWrapper-29 MuiInputBase-fullWidth-77 MuiInput-fullWidth-62 MuiInputBase-formControl-69 MuiInput-formControl-56"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input-78 MuiInput-input-63 FormInput-textFieldInput-30"
              data-testid="form-name-input"
              id="name"
              name="name"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="MuiFormControl-root-37"
        >
          <label
            class="MuiFormLabel-root-48 FormInput-textFieldLabel-32 FormInput-label-33 FormInput-singleLabel-34 MuiFormLabel-required-53"
          >
            Upload Form Base
            <span
              class="MuiFormLabel-asterisk-54"
            >
               *
            </span>
          </label>
        </div>
        <div
          class="UploadFormC-dropzoneContainer-90"
          data-testid="uploadForm"
        >
          <div
            aria-disabled="false"
            class="Dropzone-root-104"
            data-testid="dropzone"
            style="position: relative;"
          >
            <div
              class="UploadFormC-dropzone-85 UploadFormC-paddingAnimationDropzoneOff-92"
            >
              <svg
                fill="none"
                height="34"
                viewBox="0 0 37 34"
                width="37"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M27.75 9.25c-.077 0-.154 0-.23.003C26.445 3.954 21.76 0 16.187 0 9.802 0 4.624 5.177 4.624 11.563c0 .462.027.922.081 1.377a7.708 7.708 0 0 0 3.002 14.81 1.542 1.542 0 1 0 0-3.083 4.625 4.625 0 0 1-.934-9.156 1.542 1.542 0 0 0 1.187-1.884 8.48 8.48 0 0 1 8.227-10.543c4.49 0 8.194 3.5 8.463 7.959.057.936.93 1.606 1.85 1.417a6.167 6.167 0 1 1 1.25 12.207 1.542 1.542 0 1 0-.001 3.083 9.25 9.25 0 0 0 0-18.5zm-8.48 22.958V20.513L21 22.242a1.542 1.542 0 1 0 2.18-2.18l-3.27-3.27a3.083 3.083 0 0 0-4.361 0l-3.27 3.27a1.542 1.542 0 0 0 2.18 2.18l1.729-1.729v11.695a1.542 1.542 0 0 0 3.083 0z"
                  fill="#7A49DF"
                  fill-rule="evenodd"
                />
              </svg>
              <span>
                Upload PDF
              </span>
            </div>
            <input
              accept="application/pdf"
              autocomplete="off"
              style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px; opacity: 0.00001; pointer-events: none;"
              type="file"
            />
          </div>
          
        </div>
        <fieldset
          class="MuiFormControl-root-37"
        >
          <div
            aria-label="status"
            class="MuiFormGroup-root-105 FormInput-statusRadioGroup-35"
            role="radiogroup"
          >
            <label
              class="MuiFormControlLabel-root-107"
              data-testid="active-form"
            >
              <span
                class="MuiButtonBase-root-14 MuiIconButton-root-8 MuiSwitchBase-root-116 MuiRadio-root-111 MuiRadio-colorSecondary-115 MuiSwitchBase-checked-117 MuiRadio-checked-112"
              >
                <span
                  class="MuiIconButton-label-13"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root-17"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zm0-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <input
                    checked=""
                    class="MuiSwitchBase-input-119"
                    name="status"
                    type="radio"
                    value="ACTIVE"
                  />
                </span>
                <span
                  class="MuiTouchRipple-root-120"
                />
              </span>
              <span
                class="MuiTypography-root-127 MuiTypography-body1-136 MuiFormControlLabel-label-110"
              >
                Active
              </span>
            </label>
            <label
              class="MuiFormControlLabel-root-107"
              data-testid="inactive-form"
            >
              <span
                class="MuiButtonBase-root-14 MuiIconButton-root-8 MuiSwitchBase-root-116 MuiRadio-root-111 MuiRadio-colorSecondary-115"
              >
                <span
                  class="MuiIconButton-label-13"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root-17"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                    />
                  </svg>
                  <input
                    class="MuiSwitchBase-input-119"
                    name="status"
                    type="radio"
                    value="INACTIVE"
                  />
                </span>
                <span
                  class="MuiTouchRipple-root-120"
                />
              </span>
              <span
                class="MuiTypography-root-127 MuiTypography-body1-136 MuiFormControlLabel-label-110"
              >
                Inactive
              </span>
            </label>
          </div>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiDialogActions-root-153 DialogCreatorInjector-dialogActions-4"
    >
      <button
        class="MuiButtonBase-root-14 MuiButton-root-159 MuiButton-contained-170 MuiButton-containedSecondary-172 ButtonC-secondary-157 MuiButton-raised-173 MuiButton-raisedSecondary-175 ButtonC-button-155"
        data-testid="close-dialog-button-bottom"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label-160"
        >
          Cancel
        </span>
      </button>
      <button
        class="MuiButtonBase-root-14 MuiButton-root-159 MuiButton-contained-170 MuiButton-containedPrimary-171 ButtonC-primary-156 MuiButton-raised-173 MuiButton-raisedPrimary-174 ButtonC-button-155"
        data-testid="action-dialog-button-bottom"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label-160"
        >
          <span>
            Start Editing
          </span>
        </span>
      </button>
    </div>
  </form>
</div>
`;
